src/eliminate_identity.cpp

```diff
@@ -0,0 +1,49 @@
+#include <migraphx/eliminate_identity.hpp>
+#include <migraphx/program.hpp>
+#include <migraphx/instruction.hpp>
+#include <migraphx/operators.hpp>
+#include <migraphx/iterator_for.hpp>
+#include <migraphx/ranges.hpp>
+#include <migraphx/stringutils.hpp>
+#include <utility>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+void eliminate_identity::apply(program& p) const
+{
+    auto last = std::prev(p.end());
+    for(auto ins : iterator_for(p))
+    {
+        // Skip the first instruction, since we always process the previous
+        // instruction
+        if(ins == p.begin())
+            continue;
+        const auto i = std::prev(ins);
+
+        if(i->name() == "identity")
+        {
+            p.replace_instruction(i, i->inputs().front());
+            p.move_instruction(i, p.end());
+        }
+        if(ins == last)
+        {
+            if(ins->name() == "identity")
+            {
+                const instruction_ref& identity_input = i->inputs().front();
+                if(identity_input->outputs().size() == 1)
+                {
+                    p.move_instruction(identity_input, i);
+                    // since this is the last instruction, removing it only
+                    // requires changing "last" and calling remove below
+                    last = std::prev(last);
+                }
+            }
```

Comment:
If identify_input->inputs().front()->outputs().size == 1, then identify_inputs->inputs().front() can also be eliminated.   There is a transitive closure here.   It is cleaner for dead code elimination pass to take care of the removal.  And the dead code elimination should  take care of transitive closure if it doesn't yet.

Reply:
I am not sure what the suggested change should be, so please correct me if I'm wrong. Is the conditional statement
`if(identity_input->outputs().size() == 1)` not necessary because dead code elimination will take care of it?

Also, we originally had inserted a dead code elimination after eliminate identity, but one of the test cases failed because the previous dependencies were being removed. Instead, we decided to just remove all identity instructions only. I will try to see if I can reproduce the failed test and its output if that will help.

Reply:
It is not safe to do since a concat could be replaced by an identity in the `eliminate_concat` pass. Even though the output of each operator is not directly used, the memory that the operators write to is used.

Reply:
[test_concat_relu](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/blob/develop/test/gpu/miopen.cpp#L1073)

State of program before eliminate_identity:
```
Pass: dead_code_elimination
@0 = check_context -> float_type, {}, {}
scratch = @param:scratch -> float_type, {12}, {1}
@2 = load[offset=0,end=48](scratch) -> float_type, {6, 2}, {2, 1}
@3 = load[offset=0,end=16](@2) -> float_type, {2, 2}, {2, 1}
x = @param:x -> float_type, {2, 2}, {2, 1}
@5 = gpu::set_stream[stream=0] -> float_type, {}, {}
@6 = gpu::relu(x,@3) -> float_type, {2, 2}, {2, 1}
@7 = load[offset=16,end=40](@2) -> float_type, {3, 2}, {2, 1}
y = @param:y -> float_type, {3, 2}, {2, 1}
@9 = gpu::relu(y,@7) -> float_type, {3, 2}, {2, 1}
@10 = load[offset=40,end=48](@2) -> float_type, {1, 2}, {2, 1}
z = @param:z -> float_type, {1, 2}, {2, 1}
@12 = gpu::relu(z,@10) -> float_type, {1, 2}, {2, 1}
@13 = identity(@2,@6,@9,@12) -> float_type, {6, 2}, {2, 1}
output = @param:output -> float_type, {6, 2}, {2, 1}
@15 = gpu::relu(@13,output) -> float_type, {6, 2}, {2, 1}
```

State of program just after eliminate_identity:
```
Pass: eliminate_identity
@0 = check_context -> float_type, {}, {}
scratch = @param:scratch -> float_type, {12}, {1}
@2 = load[offset=0,end=48](scratch) -> float_type, {6, 2}, {2, 1}
@3 = load[offset=0,end=16](@2) -> float_type, {2, 2}, {2, 1}
x = @param:x -> float_type, {2, 2}, {2, 1}
@5 = gpu::set_stream[stream=0] -> float_type, {}, {}
@6 = gpu::relu(x,@3) -> float_type, {2, 2}, {2, 1}
@7 = load[offset=16,end=40](@2) -> float_type, {3, 2}, {2, 1}
y = @param:y -> float_type, {3, 2}, {2, 1}
@9 = gpu::relu(y,@7) -> float_type, {3, 2}, {2, 1}
@10 = load[offset=40,end=48](@2) -> float_type, {1, 2}, {2, 1}
z = @param:z -> float_type, {1, 2}, {2, 1}
@12 = gpu::relu(z,@10) -> float_type, {1, 2}, {2, 1}
output = @param:output -> float_type, {6, 2}, {2, 1}
@14 = gpu::relu(@2,output) -> float_type, {6, 2}, {2, 1}
```

State of program after final dead_code_elimination:
```
Pass: dead_code_elimination
@0 = check_context -> float_type, {}, {}
scratch = @param:scratch -> float_type, {12}, {1}
@2 = load[offset=0,end=48](scratch) -> float_type, {6, 2}, {2, 1}
@3 = gpu::set_stream[stream=0] -> float_type, {}, {}
output = @param:output -> float_type, {6, 2}, {2, 1}
@5 = gpu::relu(@2,output) -> float_type, {6, 2}, {2, 1}

FAILED: test_concat_relu
error: 0.309295
cpu:0.5, 0.75, 0, 0.6875, 0.0625, 0.75, 0, 0.125, 0, 0, 0, 0.75
gpu:0.8125, 0.75, -0, 0.875, -0, 0.4375, -0, 0.25, -0, 0.375, 0.6875, 0.875
Max diff: 0.6875
Mismatch at 0: 0.5 != 0.8125

z = @param:z -> float_type, {1, 2}, {2, 1}
y = @param:y -> float_type, {3, 2}, {2, 1}
x = @param:x -> float_type, {2, 2}, {2, 1}
@3 = relu(x) -> float_type, {2, 2}, {2, 1}
@4 = relu(y) -> float_type, {3, 2}, {2, 1}
@5 = relu(z) -> float_type, {1, 2}, {2, 1}
@6 = concat(@3,@4,@5) -> float_type, {6, 2}, {2, 1}
@7 = relu(@6) -> float_type, {6, 2}, {2, 1}

cpu:
z = @param:z -> float_type, {1, 2}, {2, 1}
y = @param:y -> float_type, {3, 2}, {2, 1}
x = @param:x -> float_type, {2, 2}, {2, 1}
@3 = cpu::relu(x) -> float_type, {2, 2}, {2, 1}
@4 = cpu::relu(y) -> float_type, {3, 2}, {2, 1}
@5 = cpu::relu(z) -> float_type, {1, 2}, {2, 1}
@6 = cpu::concat(@3,@4,@5) -> float_type, {6, 2}, {2, 1}
@7 = cpu::relu(@6) -> float_type, {6, 2}, {2, 1}

gpu:
@0 = check_context -> float_type, {}, {}
scratch = @param:scratch -> float_type, {12}, {1}
@2 = load[offset=0,end=48](scratch) -> float_type, {6, 2}, {2, 1}
@3 = gpu::set_stream[stream=0] -> float_type, {}, {}
output = @param:output -> float_type, {6, 2}, {2, 1}
@5 = gpu::relu(@2,output) -> float_type, {6, 2}, {2, 1}
```

Reply:
I would like to note that using dead code elimination after eliminate identity is okay provided that the program's overall state is not affected (the example above shows where this is not the case). I can explain this behavior in the code, assuming everything else is understood so far.